Utilities/XrdAdaptor/src/XrdStatistics.cc

```diff
@@ -0,0 +1,180 @@
+    {
+        XrdSiteStatisticsInformation::createInstance();
+    }
+
+    iRegistry.watchPostEndJob(this, &XrdStatisticsService::postEndJob);
+}
+
+
+void XrdStatisticsService::postEndJob()
+{
+    edm::Service<edm::JobReport> reportSvc;
+    if (!reportSvc.isAvailable()) {return;}
+
+    XrdSiteStatisticsInformation *instance = XrdSiteStatisticsInformation::getInstance();
+    if (!instance) {return;}
+
+    std::map<std::string, std::string> props;
+    for (std::shared_ptr<XrdSiteStatistics> const &stats : instance->m_sites)
+    {
+        stats->recomputeProperties(props);
+        reportSvc->reportPerformanceForModule(stats->site(), "XrdSiteStatistics", props);
+    }
+}
+
+
+std::shared_ptr<XrdSiteStatistics>
+XrdSiteStatisticsInformation::getStatisticsForSite(std::string const &site)
+{
+    std::lock_guard<std::mutex> lock(m_mutex);
+    for (std::shared_ptr<XrdSiteStatistics> &stats : m_sites)
+    {
+        if (stats->site() == site) {return stats;}
+    }
+    m_sites.emplace_back(new XrdSiteStatistics(site));
+    return m_sites.back();
+}
+
+
+void
+XrdSiteStatisticsInformation::createInstance()
+{
+    if (!m_instance)
+    {
+        std::unique_ptr<XrdSiteStatisticsInformation> tmp { new XrdSiteStatisticsInformation() };
+        XrdSiteStatisticsInformation* expected = nullptr;
+        if (m_instance.compare_exchange_strong(expected,tmp.get()))
+        {
+            tmp.release();
+        }
+    }
+}
+
+XrdSiteStatisticsInformation *
+XrdSiteStatisticsInformation::getInstance()
+{
+    return m_instance.load(std::memory_order_relaxed);
+}
+
+void XrdStatisticsService::fillDescriptions(edm::ConfigurationDescriptions &descriptions)
+{
+    edm::ParameterSetDescription desc;
+    desc.setComment("Report Xrootd-related statistics centrally.");
+    desc.addUntracked<bool>("reportToFJR", true)
+        ->setComment("True: Add per-site Xrootd statistics to the framework job report.\n"
+                     "False: Collect no site-specific statistics.\n");
+    descriptions.add("XrdAdaptor::XrdStatisticsService", desc);
+}
+
+
+XrdSiteStatistics::XrdSiteStatistics(std::string const &site) :
+    m_site(site),
+    m_readvCount(0),
+    m_chunkCount(0),
+    m_readvSize(0),
+    m_readvNS(0.0),
+    m_readCount(0),
+    m_readSize(0),
+    m_readNS(0)
+{
+}
+
+std::shared_ptr<XrdReadStatistics>
+XrdSiteStatistics::startRead(std::shared_ptr<XrdSiteStatistics> parent, std::shared_ptr<ClientRequest> req)
+{
+    std::shared_ptr<XrdReadStatistics> readStats(new XrdReadStatistics(parent, req->getSize(), req->getCount()));
+    return readStats;
+}
+
+
+static std::string
+i2str(int input)
+{
+    std::ostringstream formatter;
+    formatter << input;
+    return formatter.str();
+}
+
+
+static std::string
+d2str(double input)
+{
+    std::ostringstream formatter;
+    formatter << std::setw(4) << input;
+    return formatter.str();
+}
+
+
+void
+XrdSiteStatistics::recomputeProperties(std::map<std::string, std::string> &props)
```

Comment:
In the future, this should really be made ```const```. I won't require it for this pull request.